drivers/interrupt_controller/intc_swi_nrf_egu.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public APIs for nrf_irq_offload subsystem
+ */
+
+#include <init.h>
+#include <kernel.h>
+#include <drivers/interrupt_controller/intc_nrf_egu.h>
+#include <drivers/interrupt_controller/intc_swi.h>
+
+struct swi_data {
+	const struct device *egu;
+	int channel;
+	sys_slist_t swi_channels;
+	struct k_spinlock lock;
+};
+
+static struct swi_data data = {
+	.swi_channels = SYS_SLIST_STATIC_INIT(&data.swi_channels)
+};
+
+static void egu_callback(const struct device *dev, int channel, void *ctx)
+{
+	sys_slist_t sig_list = SYS_SLIST_STATIC_INIT(&sig_list);
+	struct swi_channel *swi;
+	k_spinlock_key_t key;
+	sys_snode_t *node;
+
+	ARG_UNUSED(ctx);
+
+	key = k_spin_lock(&data.lock);
+	SYS_SLIST_FOR_EACH_CONTAINER(&data.swi_channels, swi, node) {
+		if (atomic_cas(&swi->signaled, 1, 0)) {
+			sys_slist_append(&sig_list, &swi->aux_node);
+		}
+	}
+	k_spin_unlock(&data.lock, key);
+
+	while ((node = sys_slist_get(&sig_list)) != NULL) {
+		swi = CONTAINER_OF(node, struct swi_channel, aux_node);
+		swi->cb(swi);
+	}
+}
+
+int swi_channel_init(struct swi_channel *swi, swi_channel_cb_t cb)
+{
+	k_spinlock_key_t key;
+	int ret = 0;
+
+	if (cb == NULL) {
+		return -EINVAL;
+	}
+
+	swi->cb = cb;
+	atomic_clear(&swi->signaled);
```

Comment:
Shouldn't this be done after it is checked that the channel is not already initialized?

Reply:
My assumption was that double init is undefined behaviour. I will figure out a way to prevent double init.

Reply:
Fixed